clippy_lints/src/redundant_clone.rs

```diff
@@ -70,317 +69,729 @@ impl<'tcx> LateLintPass<'tcx> for RedundantClone {
+                    statement_index: data.statements.len(),
+                });
+                if let TerminatorKind::Call { args, .. } = &data.terminator().kind
+                    && let [Operand::Move(arg), ..] = &**args
+                    && let Some(Some(targets)) = ref_target_results.get().targets.get(&arg.local)
+                {
+                    Some((block, (arg.local, targets.clone())))
+                } else {
+                    None
+                }
+            })
+            .collect::<FxHashMap<_, _>>();
+        if clone_sources.is_empty() {
+            return;
+        }
 
-        for (bb, bbdata) in mir.basic_blocks.iter_enumerated() {
-            let terminator = bbdata.terminator();
+        let mut results = CloneAnalysis::new(cx, body, &raw_ptr_taken, &clone_sources)
+            .into_engine(cx.tcx, body)
+            .iterate_to_fixpoint();
+
+        // Check if a merged values are borrowed at the start of each block.
+        // Workaround for `join` not having access to the analysis.
+        let mut required_clones = take(&mut results.analysis.required_clones);
+        let mut borrowers = PossibleBorrowerMap::new(cx, body);
+        for block in (0..body.basic_blocks.len()).map(BasicBlock::from_usize) {
+            let state = results.entry_set_for_block(block);
+            let loc = block.start_location();
+            for (&local, _) in state.values.iter().filter(|(_, v)| v.last_modified.is_none()) {
+                for &(l1, l2, block) in &state.links {
+                    let local = if l1 == local {
+                        l2
+                    } else if l2 == local {
+                        l1
+                    } else {
+                        continue;
+                    };
 
-            if terminator.source_info.span.from_expansion() {
-                continue;
+                    if !borrowers.bounded_borrowers(&[], &[], local, loc) {
+                        required_clones.insert(block);
+                    }
+                }
             }
+        }
 
-            // Give up on loops
-            if terminator.successors().any(|s| s == bb) {
+        let mut infos: Vec<_> = clone_sources
+            .keys()
+            .filter(|b| !required_clones.contains(b))
+            .map(|&b| &body.basic_blocks[b].terminator().source_info)
+            .collect();
+        infos.sort_by_key(|i| i.span);
+        for info in infos {
+            let node = body.source_scopes[info.scope]
+                .local_data
+                .as_ref()
+                .assert_crate_local()
+                .lint_root;
+            span_lint_hir(cx, REDUNDANT_CLONE, node, info.span, "redundant clone");
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct Value {
+    // Two values which were last modified at different locations may contain different values.
+    // Joining two such values needs to be treated as a modification.
+    last_modified: Option<Location>,
```

Comment:
This really should be something like:
```rust
enum Index {
    Statement(usize),
    Merge(Local),
}
struct Location {
    block: BasicBlock,
    index: Index,
}
```

This would allow distinguishing different merged values from each other, but would require `join` to know which block it's working on.

Reply:
I would represent a value with a simple index. A first pass collects value roots. Those would be the value of a variable where it is borrowed before cloning.

The state would be something like `IndexVec<Local, FlatSet<Value>>`.

```rust
let mut x = if _ { foo() } else { bar() };
let y = if _ { foo2() } else { bar2() };

// `x` is borrowed here. `x` is `FlatSet::Elem(Value(0))` here.
let z = x.clone();
// `z` is `FlatSet::Elem(Value(0))`.

// assuming the match doesn't modify `x`, `y`, and `z`, then the join when merging
// the arms shouldn't detect a modification
match _ {
    _ => { ... },
    _ => { ... },
    _ => { ... },
};
// `x` and `z` are still both `FlatSet::Elem(Value(0))`

// here the clone isn't needed: `x` and `y` are both `FlatSet::Elem(Value(0))`.
do_something(&z);

if _ {
    x = y;
    // `x` would be `FlatSet::Top` if `y` is not tracked.
}
// By joining, `x` is `FlatSet::Top`.

// here the clone is needed: `x` is `FlatSet::Top`, while `z` is `FlatSet::Elem(Value(0))`
do_something(&z);
```

Reply:
That wouldn't handle cloning a value twice correctly, nor would it handle a chain of clones. e.g.
```rust
let x = foo();
let y = if bar {
    x.clone()
} else {
    x.clone()
};
let z = y.clone();
// `x`, `y` and `z` should all be seen as the same value here.
```